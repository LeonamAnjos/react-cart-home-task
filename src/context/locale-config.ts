export type LocaleType = "de-DE" | "en-US";

type LocaleConfig = {
  currency: string;
};

export const localeMap: Record<LocaleType, LocaleConfig> = {
  "de-DE": { currency: "EUR" },
  "en-US": { currency: "USD" },
};
