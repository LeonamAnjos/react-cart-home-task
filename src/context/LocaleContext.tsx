import React, { createContext, ReactNode, useMemo, useState } from "react";
import { localeMap, LocaleType } from "./locale-config";

type LocaleState = {
  getLocale: () => LocaleType | undefined;
  setLocale: (locale: LocaleType) => void;
  currencyFormat: (num: number) => string;
};

const LocaleContext = createContext<LocaleState>({
  getLocale: (): LocaleType | undefined => undefined,
  setLocale: (locale: LocaleType): void => {
    /* EMPTY */
  },
  currencyFormat: (num: number): string => "",
});

export interface LocaleProviderProps {
  children?: ReactNode;
  locale: LocaleType;
}

export const LocaleProvider = ({ children, locale }: LocaleProviderProps) => {
  const [value, setValue] = useState(locale);

  const contextValue = useMemo(() => {
    const formatter = new Intl.NumberFormat(value, {
      style: "currency",
      currency: localeMap[value]?.currency,
      minimumFractionDigits: 2,
    });

    return {
      getLocale: () => value,
      setLocale: (locale: LocaleType) => setValue(locale),
      currencyFormat: (num: number): string => formatter.format(num),
    };
  }, [value, setValue]);

  return (
    <LocaleContext.Provider value={contextValue}>
      {children}
    </LocaleContext.Provider>
  );
};

export default LocaleContext;
