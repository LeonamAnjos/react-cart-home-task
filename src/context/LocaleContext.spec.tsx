import { render } from "@testing-library/react";
import LocaleContext, { LocaleProvider } from "./LocaleContext";
import { LocaleType } from "./locale-config";
import { ReactNode } from "react";

const customRender = (children: ReactNode, locale: LocaleType) => {
  return render(<LocaleProvider locale={locale}>{children}</LocaleProvider>);
};

describe("LocaleContext", () => {
  it("WHEN rendered", () => {
    const { container } = customRender("", "de-DE");

    expect(container).toMatchSnapshot();
  });
});
