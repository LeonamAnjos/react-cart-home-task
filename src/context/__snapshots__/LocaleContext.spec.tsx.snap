// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocaleContext WHEN rendered 1`] = `<div />`;
